### ISSUE TITLE
cs2cs fails to transform ESRI:31491 coordinates with BETA2007.gsb grid

### ISSUE THREAD

**Issue by don-vip** [OSGeo/PROJ#2508]
#### Example of problem

I try to update the [ESRI proj-string definitions](https://josm.openstreetmap.de/browser/josm/trunk/nodist/data/projection/esri) we ship in JOSM. The update process involves coordinates transformations using `cs2cs` to make sure JOSM provides the same results as PROJ.

ESRI [changed](https://github.com/Esri/projection-engine-db-doc/pull/14/files?file-filters%5B%5D=No+extension#diff-0e8e4615452f81489dd7fe120d6a3ddf103b6d9ca71cb8b2d9d61d721ab3517cR1383) its definition of ESRI:31491 to add `+nadgrids=BETA2007.gsb`:

```diff
 # Germany_Zone_1 [Germany Zone 1]
 # area: (lat: 55.24, 55.92) - (lon: 3.34, 4.5) [Germany - offshore North Sea west of 4.5~E]
-<31491> +proj=tmerc +ellps=bessel +x_0=1500000.0 +y_0=0.0 +lon_0=3dE +lat_0=0dN +k_0=1.0 +no_defs <>
+<31491> +proj=tmerc +ellps=bessel +x_0=1500000.0 +y_0=0.0 +lon_0=3dE +lat_0=0dN +k_0=1.0 +nadgrids=BETA2007.gsb +no_defs <>
```

I'm unable to perform coordinates transformations with `+nadgrids`:
```sh
$ PROJ_DEBUG=2 cs2cs -f %.9f +proj=longlat +datum=WGS84 +to +proj=tmerc +ellps=bessel +x_0=1500000.0 +y_0=0.0 +lon_0=3dE +lat_0=0dN +k_0=1.0 +nadgrids=BETA2007.gsb +no_defs +bounds=3.34,55.24,4.5,55.92
pj_open_lib(proj.db): call fopen(/home/.../proj.db) - succeeded
pj_open_lib(BETA2007.gsb): call fopen(/home/.../BETA2007.gsb) - succeeded
pj_open_lib(de_adv_BETA2007.tif): call fopen(/home/.../de_adv_BETA2007.tif) - failed
pj_open_lib(BETA2007.gsb): call fopen(/home/.../BETA2007.gsb) - succeeded
3.79239684537 55.77935322422
*       * inf
4.37731978264 55.87383807976
*       * inf
```

I tried to download GeoTIFF grid from your CDN, same results:
```sh
$ PROJ_DEBUG=2 cs2cs -f %.9f +proj=longlat +datum=WGS84 +to +proj=tmerc +ellps=bessel +x_0=1500000.0 +y_0=0.0 +lon_0=3dE +lat_0=0dN +k_0=1.0 +nadgrids=BETA2007.gsb +no_defs +bounds=3.34,55.24,4.5,55.92
pj_open_lib(proj.db): call fopen(/home/.../proj.db) - succeeded
pj_open_lib(BETA2007.gsb): call fopen(/home/.../BETA2007.gsb) - succeeded
pj_open_lib(de_adv_BETA2007.tif): call fopen(/home/.../de_adv_BETA2007.tif) - succeeded
3.79239684537 55.77935322422
*       * inf
4.37731978264 55.87383807976
*       * inf
```

#### Problem description

cs2cs fails to transform coordinates (`*       * inf`) using the BETA2007 grid.

#### Expected Output

Something similar to output without `+nadgrids`:
```sh
$ PROJ_DEBUG=2 cs2cs -f %.9f +proj=longlat +datum=WGS84 +to +proj=tmerc +ellps=bessel +x_0=1500000.0 +y_0=0.0 +lon_0=3dE +lat_0=0dN +k_0=1.0 +no_defs +bounds=3.34,55.24,4.5,55.92
pj_open_lib(proj.db): call fopen(/home/.../proj.db) - succeeded
3.79239684537 55.77935322422
1549714.449533752       6183639.034067011 0.000000000
4.37731978264 55.87383807976
1586200.866999898       6194731.100542467 0.000000000
```

#### Environment Information
 - PROJ 7.2.1
 - Ubuntu 18.04.5 LTS

#### Installation method
 - from source

**Comment by metzm** [comment]
The failing coordinates are not covered by the BETA2007 grid. You need to make sure that the coordinates you want to transform are actually covered by the datum shift grid. If not, don't use the grid.

**Comment by metzm** [comment]
The cs2cs tool is legacy, it seems that cct is now the recommended tool for coordinate conversion and transformation. With cct 
```
echo "3.79239684537 55.77935322422" | cct -z0 -t0 +proj=longlat +datum=WGS84 +to +proj=tmerc +ellps=bessel +x_0=1500000.0 +y_0=0.0 +lon_0=3dE +lat_0=0dN +k_0=1.0 +nadgrids=BETA2007.gsb +no_defs +bounds=3.34,55.24,4.5,55.92
```
I get
```
# Record 0 TRANSFORMATION ERROR: 3.79239684537 55.77935322422
 (point not within available datum shift grids)
```
The extents of the BETA2007 grid (as of PROJ7 this is "de_adv_BETA2007.tif" as mentioned in your cs2cs output) are reported by gdalinfo as
```
Upper Left  (   5.4166667,  55.3500000) (  5d25' 0.00"E, 55d21' 0.00"N)
Lower Right (  15.7500000,  46.9500000) ( 15d45' 0.00"E, 46d57' 0.00"N)
```

**Comment by don-vip** [comment]
OK! Thank you @metzm for explaining the reason to me :) I'll switch to cct and ask ESRI directly why they added the grid to this particular projection.

**Comment by kbevers** [comment]
It seems that the core issue has been resolved, here's just a few clarifications.

>  The cs2cs tool is legacy, it seems that cct is now the recommended tool for coordinate conversion and transformation.

`cs2cs` is a part of the PROJ legacy, yes. But it is not a legacy tool in the sense that its use is discouraged. Quite the contrary. `cs2cs` is the primary tool to transform between one CRS and another. Proj-strings as CRS identifiers are deprecated and highly discouraged though. Use `cs2cs` with URN's for best results (e.g. `cs2cs EPSG:xxxx EPSG:yyyy`).

> `... cct -z0 -t0 +proj=longlat +datum=WGS84 +to +proj=tmerc +ellps=bessel +x_ ...`

The `+to` parameter is ignored by `cct` and it will NOT make `cct` behave like `cs2cs`.

In short, `cs2cs` is used to transform between to named coordinate reference systems. `cct` is the power users tool that will process coordinates with whatever pipeline it is served.

**Comment by metzm** [comment]
> > `... cct -z0 -t0 +proj=longlat +datum=WGS84 +to +proj=tmerc +ellps=bessel +x_ ...`
> 
> The `+to` parameter is ignored by `cct` and it will NOT make `cct` behave like `cs2cs`.

Oops, sorry for the sloppy cut'n paste!

> 
> In short, `cs2cs` is used to transform between to named coordinate reference systems. `cct` is the power users tool that will process coordinates with whatever pipeline it is served.

In this case I prefer cct because contrary to cs2cs it checks for and reports transformation errors. That made me guess that cs2cs is legacy.

**Comment by metzm** [comment]
Back to this grid. IIUC, adding `+nadgrids=BETA2007.gsb` to ESRI:31491 creates a bound crs that is restricted to the use case of this grid: only datum transformations between DHDN and ETRS89, and only for the area covered by this grid, east from 5d25' E to 15d45' E. However, ESRI:31491 is defined to use Gauss-Kruger zone 1, by definition going east from 1d20' E to 4d40' E (small overlap between G-K zones), thus not even overlapping with the BETA2007 grid. Even with a relaxed coverage of 6 degrees easting (0d E to 6d E) in ESRI:31491, there would be only a small overlap at the western edge of the grid. Thus adding `+nadgrids=BETA2007.gsb` to ESRI:31491 seems to be a mistake.
